### PR TITLE
Shenli/string literal

### PR DESCRIPTION
**Pull request by shenli**
Parse sql text stringLiteral to DataItem.

### CODE REVIEW

File: expression/like.go [hunk 1]
```diff
@@ -107,6 +107,8 @@ func (p *PatternLike) Eval(ctx context.Context, args map[interface{}]interface{}
 			spattern = v
 		case []byte:
 			spattern = string(v)
+		case *types.DataItem:
+			spattern = v.Data.(string)
```
Comment (siddontang) [inline expression/like.go]: check v.Data type too.

File: parser/coldef/col_def.go [hunk 2]
```diff
@@ -103,8 +103,11 @@ func getDefaultValue(c *ConstraintOpt, tp byte, fsp int) (interface{}, error) {
 
 		return value, nil
 	}
-
-	return expression.FastEval(c.Evalue), nil
+	v := expression.FastEval(c.Evalue)
+	if vv, ok := v.(*types.DataItem); ok {
```
Comment (siddontang) [inline parser/coldef/col_def.go]: put this in FastEval?

Comment (shenli) [inline parser/coldef/col_def.go]: FastEval is used somewhere else. I am not sure about this. I am concerned  about the side effects.

File: expression/like.go (hunk 1 above)
Comment (shenli) [inline expression/like.go]: PTAL

File: parser/coldef/col_def.go (hunk 2 above)
Comment (siddontang) [inline parser/coldef/col_def.go]: maybe abstract a function naming IsDateItem may be better for other places check.

Comment (shenli) [inline parser/coldef/col_def.go]: PTAL @siddontang

**Comment by siddontang** [comment]
LGTM

File: expression/like.go [hunk 3]
```diff
@@ -101,14 +101,10 @@ func (p *PatternLike) Eval(ctx context.Context, args map[interface{}]interface{}
...
-		case string:
-			spattern = v
-		case []byte:
-			spattern = string(v)
-		default:
-			return nil, errors.Errorf("Pattern should be string or []byte in LIKE: %v (Value of type %T)", pattern, pattern)
+		spattern, err := types.ToString(pattern)
+		if err != nil {
+			return nil, errors.Trace(err)
+			//errors.Errorf("Pattern should be string or []byte in LIKE: %v (Value of type %T)", pattern, pattern)
```
Comment (zimulala) [inline expression/like.go]: need a space？

Comment (shenli) [inline expression/like.go]: You mean remove the comment?

File: expression/value.go [hunk 4]
```diff
@@ -45,14 +47,19 @@ func (l Value) IsStatic() bool {
...
 		return fmt.Sprintf("%q", x)
-	default:
-		return fmt.Sprintf("%v", l.Val)
+	case *types.DataItem:
+		if x.Type.Tp == mysql.TypeString {
+			return fmt.Sprintf("%q", x.Data)
+		}
+		return fmt.Sprintf("%v", x.Data)
 	}
+	return fmt.Sprintf("%v", l.Val)
```
Comment (qiuyesuifeng) [inline expression/value.go]: Why not using default in switch as old code?

Comment (qiuyesuifeng) [inline expression/value.go]: And also `fmt.Sprintf("%v", x)`?

Comment (shenli) [inline expression/value.go]: For string literal, we need to print a string with quotation mark, so we need use %q.

File: util/types/field_type.go [hunk 5]
```diff
@@ -136,3 +136,9 @@ func RawData(d interface{}) interface{} {
 func IsNil(d interface{}) bool {
 	return RawData(d) == nil
 }
+
+// IsDataItem checks if the interface is a DataItem.
+func IsDataItem(d interface{}) bool {
```
Comment (qiuyesuifeng) [inline util/types/field_type.go]: Seems useless?

Comment (shenli) [inline util/types/field_type.go]: It will be used latter.

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**